### PR TITLE
feat: insert markdown value in the current editor selection point.

### PR DESCRIPTION
**Pull request by forhappy**
This is to support inserting markdown content in the current selection(where the cursor is) point, we can programmatically insert new markdown content to the existing markdown editor now.

NOTE: the Signal `insertMarkdown$` can be only used in plugins, there is no way to use this utility API from outside of the editor.

https://github.com/mdx-editor/editor/assets/706422/24fcd570-ab4e-4940-92ca-f199ad9ee143

### New idea to discuss:
We can even expose a MDXEditor API(`insertMarkdown`) so that the user can insert markdown content from outside of the editor.

### CODE REVIEW

**Comment by forhappy** [comment]
This is a new version of markdown insert feature based on MDXEditor V2, happy to discuss the implementation ideas ;) 

Hey @petyosi would you help take a look into this PR?

File: src/importMarkdownToLexical.ts [hunk 1]
```diff
@@ -79,7 +79,7 @@ function isParent(node: unknown): node is Mdast.Parent {
  * @internal
  */
 export interface MdastTreeImportOptions {
-  root: LexicalRootNode
+  root: LexicalNode
```
Comment (forhappy) [inline src/importMarkdownToLexical.ts:82]: `LexicalRootNode` is a child class of `LexicalNode`, it's OK to use `LexicalNode` here as we need to insert the new markdown value into a LexicalNode.

**Comment by petyosi** [comment]
Hey @forhappy , 

I mentioned that this should be a built-in feature; here's what I have in mind as a possible implementation: https://github.com/mdx-editor/editor/commit/ad5a516241fa056abc5f7fe58ef2c2422e777f3a. 

This eliminates the need of an additional mdast import point node, which I see as a potential problem down the road. 

The commit needs some clean up of course, as I'm just butchering the root type, but you will get the idea. Does this work for you? Thank you very much for your help!

**Comment by forhappy** [comment]
Hey @petyosi Thanks for your comments and sample code, your solution is simpler and easy to understand, let me verify if it would work in some edge cases I have observed, will let you know if everything works well.

A side 🗒️: while I am happy to rewrite my PR based on your idea, I don't meant to take your credit:) so it's totally ok for you to refine your idea and polish  your change then I will cancel my PR. But if you don't mind I can raise a second revision and finalize this feature with you.

**Comment by petyosi** [comment]
That would be amazing! Please go ahead and take that approach in your PR. It is great that you have some edge cases to test it with. 

Cheers,

**Comment by braincore** [comment]
Will `insertMarkdown` be the recommended way to add something like a user-mentions dialog?

**Comment by forhappy** [comment]
> Will `insertMarkdown` be the recommended way to add something like a user-mentions dialog?

It can be used to insert markdown content(in the future we can even insert HTML content) to the current cursor point, one potential use case is the user can now input markdown content from somewhere(not in the editor) and then insert the markdown to the editor. This will be very helpful if you build something like Notion AI where the content is generated by AI and the user would like to insert the AI generated content into the editor.

**Comment by forhappy** [comment]
Hey @petyosi , I've updated the implementations and examples based on your suggestions/ideas, I've done very minor changes on the original `fakeRoot`(added comments and removed the type conversion by `new RootNode` directly), besides I added one example where `NestedEditor` is used via JSX component.

I also tested a few edge cases and it works(e.g. insert markdown content in a table cell), thank you so much for the guidance ;)

File: src/plugins/core/index.ts [hunk 2]
```diff
@@ -304,6 +305,36 @@ export const setMarkdown$ = Signal<string>((r) => {
...
+        const rootNodes: LexicalNode[] = []
+        // The shadow root is not the actual editor RootNode, it's more like a placeholder root
+        // that will be used to import markdown content because 'tryImportingMarkdown' utility
+        // will import and convert markdown content to Mdast tree first and then convert to lexical nodes,
+        // during which process we need a root lexical node to host all children nodes
+        // while we're visiting the Mdast tree nodes.
+        const shadowRoot: RootNode = new RootNode()
+        shadowRoot.append(...rootNodes)
+        tryImportingMarkdown(r, shadowRoot, markdownToInsert)
+        $insertNodes(rootNodes)
```
[the remark this answers is not in the document]
Comment (forhappy) [reply src/plugins/core/index.ts:333]: Yes, I oversight this issue, I have changed back to use the original code and it now works. Thank you for your help.

**Review by forhappy — COMMENTED** [review]
Updated the insert logic and now it works

File: src/plugins/core/index.ts [hunk 3]
```diff
@@ -304,6 +305,36 @@ export const setMarkdown$ = Signal<string>((r) => {
...
+      const selection = $getSelection()
+      if (selection !== null) {
+        const rootNodes: LexicalNode[] = []
+        // The shadow root is not the actual editor RootNode, it's more like a placeholder root
+        // that will be used to import markdown content because 'tryImportingMarkdown' utility
+        // will import and convert markdown content to Mdast tree first and then convert to lexical nodes,
+        // during which process we need a root lexical node to host all children nodes
+        // while we're visiting the Mdast tree nodes.
+        const shadowRoot: RootNode = new RootNode()
+        shadowRoot.append(...rootNodes)
```
[the remark this answers is not in the document]
Comment (forhappy) [reply src/plugins/core/index.ts]: You're right, new a RootNode won't work, I could have missed some steps to verify it, now I changed back to use your code and it can work.
